akka-stream/src/main/scala/akka/stream/impl/fusing/CompleteOn.scala

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright (C) 2022 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream.impl.fusing
+import scala.util.control.NonFatal
+
+import akka.annotation.InternalApi
+import akka.stream.Attributes
+import akka.stream.Attributes.SourceLocation
+import akka.stream.impl.Stages.DefaultAttributes
+import akka.stream.impl.fusing.GraphStages.SimpleLinearGraphStage
+import akka.stream.stage.GraphStageLogic
+import akka.stream.stage.InHandler
+import akka.stream.stage.OutHandler
+
+@InternalApi private[akka] class CompleteOn[T](val pf: PartialFunction[Throwable, Boolean])
+    extends SimpleLinearGraphStage[T] {
+
+  override protected def initialAttributes: Attributes = DefaultAttributes.completeOn and SourceLocation.forLambda(pf)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new GraphStageLogic(shape) with InHandler with OutHandler {
+      import Collect.NotApplied
+
+      override def onPush(): Unit = push(out, grab(in))
+
+      override def onUpstreamFailure(ex: Throwable): Unit =
+        try pf.applyOrElse(ex, NotApplied) match {
+          case true => completeStage()
+          case _    => failStage(ex)
+        } catch {
+          case NonFatal(ex) => failStage(ex)
+        }
+
+      override def onPull(): Unit = pull(in)
+
+      setHandlers(in, out, this)
+    }
+}
```

Comment:
I think we should do with existing operators `recoverWith { ... => Stream.empty }` instead of a custom operator for this.

Reply:
Yes, that's my previous version, but which will need a `pf.andThen` seems a little complicate too.

Reply:
Why do we need `andThen` for that?

Reply:
When use `recoverWith` will be :
```scala
(pf:[Throwable, Boolean]).andThen({
case true => Source.empty
}
```